### PR TITLE
Passing multiple global IDs to local library holdings service

### PR DESCRIPTION
**Pull request by natlibesa**
Preliminary draft version for testing new features.

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
There seems to be multiple features added in this pull request. Please separate different features into multiple pull requests (This pull request should contain only the features for passing multiple global IDs to local library holdings services).

Also, you can remove the `##################  2017 IX` lines. They are not needed when the code is in version control and the changes can be diffed.

File: perl/holdings_proxy.cgi [hunk 1]
```diff
@@ -33,7 +33,7 @@ use Cwd 'abs_path';
 use File::Basename qw(dirname);
 
 my $cmd_path = dirname(abs_path($0));
-my $config_ref = do("$cmd_path/holdings_proxy.config");
+my $config_ref = do("$cmd_path/holdings_proxy_esa.config"); # HUOM.
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: The path to the config file must be `"$cmd_path/holdings_proxy.config")` in version control. You can change the path on the server when testing.

File: perl/holdings_proxy.cgi [hunk 2]
```diff
@@ -2,7 +2,7 @@
 #!/usr/bin/perl
 #
 # Voyager XML holdings proxy -> JSON
-# Copyright (c) 2015-2017 University Of Helsinki (The National Library Of Finland)
+# Copyright (c) 2015 University Of Helsinki (The National Library Of Finland)
```
Comment (ghost) [inline perl/holdings_proxy.cgi:5]: Fix copyright years from `2015` to `2015-2017`

File: perl/holdings_proxy.cgi [hunk 3]
```diff
@@ -86,6 +86,51 @@ my $g_callback = '';
       }
     }
   }
+
+  ##################  2017 IX ################  MELINDA-193
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy.cgi [hunk 4]
```diff
@@ -105,7 +150,18 @@ my $g_callback = '';
     }
   }
   $url .= 'globalBibId=' . url_encode($id);
-  
+
+################# 2017 IX -B-  #################  -> MELINDA-193
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy.cgi [hunk 5]
```diff
@@ -105,7 +150,18 @@ my $g_callback = '';
...
+$testi .= " \n url (original): " . $url ;
+  foreach my $kidsrow (@kids) {
+  	chop($kidsrow);
+	$url = $url .  "&globalBibId=" . $kidsrow ;
+  }
+
+$testi .= " \n url: (added) " . $url ;
+
+################# 2017 IX -B-  #################  <- MELINDA-193
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

**Review by ghost — CHANGES_REQUESTED** [review]

File: perl/holdings_proxy.cgi [hunk 6]
```diff
@@ -86,6 +86,51 @@ my $g_callback = '';
...
+      {
+           $testi = $testi . "\n"  . $field->{'code'}  ;
+           $testi = $testi . "= " .  $field->{'data'}  ;
+        my $found = $field->{'data'}  ;
+           $found =~ s/^....\(FI-MELINDA\)//g ;
+        push(@kids, $found);
+      }
+
+      # get from DATA where there is Melinda number 
+      if ($field->{'data'} =~ /^....\(FI-MELINDA\).*/ ) {
```
Comment (natlibfi-arlehiko) [inline perl/holdings_proxy.cgi]: 035 is a variable data field, so you need to loop the z-subfield to check for the string _FI-MELINDA_

File: perl/getHoldings.cgi [hunk 7]
```diff
@@ -0,0 +1,674 @@
+#!/opt/CSCperl/current/bin/perl -w
```
Comment (ghost) [inline perl/getHoldings.cgi]: This file should be in separate PR for natlibfi/voyager-holdings repository

File: perl/holdings_proxy.cgi [hunk 8]
```diff
@@ -22,7 +22,7 @@
 #
 # Required parameters when called via cgi:
 # id   the 001 of the source record
-# lib  Aleph owning library ID
+# lib  Aleph owning library ID  / v.2017.12.1
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Versioning inline is not necessary (The source control repository provides this info)

File: perl/holdings_proxy.cgi [hunk 9]
```diff
@@ -33,7 +33,7 @@ use Cwd 'abs_path';
 use File::Basename qw(dirname);
 
 my $cmd_path = dirname(abs_path($0));
-my $config_ref = do("$cmd_path/holdings_proxy.config");
+my $config_ref = do("$cmd_path/holdings_proxy_esa.config");         # HUOM. !!!
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Configuration path must be _$cmd_path/holdings_proxy.config"_

File: perl/holdings_proxy.cgi [hunk 10]
```diff
@@ -86,6 +86,39 @@ my $g_callback = '';
...
+      if ($field->{'data'} =~ /^....\(FI-MELINDA\).*/ ) {             
+        my $found = $field->{'data'}  ;
+           $found =~ s/^....\(FI-MELINDA\)//g ; 
+       push(@kids, $found); 
+      }
+
+   } # foreach my $field <-
+
+
+  ################# 2017 IX  #################  <- MELINDA-193
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy.cgi [hunk 11]
```diff
@@ -104,8 +137,30 @@ my $g_callback = '';
       $url .= '?';
     }
   }
-  $url .= 'globalBibId=' . url_encode($id);
+
+   push(@kids, url_encode($id) );
+
+################# 2017 IX -B-  #################  -> MELINDA-193
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy.cgi [hunk 12]
```diff
@@ -104,8 +137,30 @@ my $g_callback = '';
...
+  	if ($previous eq $kidsrow) { }  # skip
+           else {
+	      $url = $url .  "&globalBibId=" . url_encode($kidsrow) ;
+              $previous = $kidsrow;
+       }
+  }
+
+$testi .= " \n url: (added_&_uniqd) " . $url ;
+
+################# 2017 IX -B-  #################  <- MELINDA-193
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy.cgi [hunk 13]
```diff
@@ -217,6 +272,18 @@ my $g_callback = '';
       $fields{'textual_holdings_stmt'} =~ s/^\s//;
     }
 
+ ######### add IX 2017 ##########
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy.cgi [hunk 14]
```diff
@@ -217,6 +272,18 @@ my $g_callback = '';
...
+
+  my @newIssues = $doc->getElementsByTagName('newIssue');     
+
+    foreach my $issue (@newIssues) {
+      my $item_enum = $issue->getAttributeNode('itemEnum')->getValue();
+       $fields{'textual_holdings_stmt'} .= "<br> - "  . "$item_enum" if ($item_enum);
+    }  
+
+ ######### add IX 2017 #########
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy.cgi [hunk 15]
```diff
@@ -293,9 +364,30 @@ my $g_callback = '';
       $locations_text .= ",\n" if ($locations_text);
       $locations_text .= "        { \"location\": \"$loc\", \"items\": \"$count\", \"available\": \"$available\", \"first_due_date\": \"$first_due\" }";
     }
+
+ ####### -> IIX & IX 2017 - ########## >
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy.cgi [hunk 16]
```diff
@@ -293,9 +364,30 @@ my $g_callback = '';
...
+        $perm_loc_extra = $location_extra->getElementsByTagName('location')->item(0)->getAttribute('dispname');
+      }
+
+      if (length($locations_text) == 0)  {
+      $locations_text= "      { \"location\": \"$perm_loc_extra\",}"; 
+      }  
+       
+   }
+ 
+####### <- IIX & IX 2017 - ######### <-
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Remove obsolete comment line

File: perl/holdings_proxy_esa_ML193.cgi [hunk 17]
```diff
@@ -0,0 +1,714 @@
+#!/opt/CSCperl/current/bin/perl
```
Comment (ghost) [inline perl/holdings_proxy_esa_ML193.cgi]: This file should not be included in the repository (Use holdings_proxy.cgi)

**Review by ghost — CHANGES_REQUESTED** [review]

File: perl/holdings_proxy.cgi [hunk 18]
```diff
@@ -381,9 +381,9 @@ $testi .= " \n url: (added_&_uniqd) " . $url ;
        
    }
  
-####### <- IIX & IX 2017 - ######### <-
 
- $locations_text = $testi;           # shows the URL  
+
+ $locations_text = $testi;
```
Comment (ghost) [inline perl/holdings_proxy.cgi]: Test-messages should not be written in production

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]
